targets/TARGET_NUVOTON/TARGET_NUC472/crypto/crypto-misc.c

```diff
@@ -207,16 +251,74 @@ bool crypto_dma_buffs_overlap(const void *in_buff, size_t in_buff_size, const vo
     return overlap;
 }
 
-static bool crypto_submodule_acquire(uint16_t *submodule_avail)
+static bool crypto_submodule_acquire(crypto_submod_sem *crypto_submod_sem, bool blocking)
 {
-    uint16_t expectedCurrentValue = 1;
-    return core_util_atomic_cas_u16(submodule_avail, &expectedCurrentValue, 0);
+    MBED_ASSERT(crypto_submod_sem);
+
+    /* Initialize semaphore only once and atomically */
+    uint16_t expectedCurrentValue = SYNCOBJ_INITSTATUS_UNINIT;
+    while (1) {
+        /* Try to initialize semaphore if it has not initialized yet */
+        if (core_util_atomic_cas_u16(&crypto_submod_sem->init_status,
```

Comment:
This model still leads to a busy-wait. I suggest you have a look at the cxa_guard implementations in mbed_retarget.cpp - uses a mutex to wait for another thread initialising. Borrow the logic (and the `singleton_lock`) from that.

Once C11 is in place you should be able to use `call_once`.

Another possibility is to use `SingletonPtr` with a custom class that contains the `Semaphore(1,1)` you need. That would mean changing this file to C++.